[tools] Add system package manger with ncurses as example

uilianries opened the pull request:
Related to https://github.com/conan-io/docs/issues/3974

Documentation PR: https://github.com/conan-io/docs/pull/3978

- Added a new section in tools/ only related to the system namespace.
- As a real example, added ncurses, it's covered natively by CMake and pkgconfig

memsharded commented on examples/tools/system/package_manager/conanfile.py:
```diff
@@ -0,0 +1,50 @@
+    package_type = "shared-library"
+    settings = "os", "arch", "compiler", "build_type"
+
+    def layout(self):
```
Can be removed

uilianries replied:
Addressed on the commit [1588288](https://github.com/conan-io/examples2/pull/177/commits/1588288ccab4c1c09e875191a15ad2669fddf2e9)

memsharded commented on examples/tools/system/package_manager/conanfile.py:
```diff
@@ -0,0 +1,50 @@
+        self.cpp_info.includedirs = []
+        self.cpp_info.libdirs = []
+
+        if self.settings.os != "Linux":
```
But the ``system_requirements()`` didn't install anything, so what would it detect?

uilianries replied:
Addressed by [1588288](https://github.com/conan-io/examples2/pull/177/commits/1588288ccab4c1c09e875191a15ad2669fddf2e9)

memsharded commented on examples/tools/system/package_manager/conanfile.py:
```diff
@@ -0,0 +1,50 @@
+        self.info.clear()
+
+    def system_requirements(self):
+        if self.settings.os != "Linux":
```
Better a validate() that raises, having this in Windows as a no-op would be very confusing

uilianries replied:
Addressed by [1588288](https://github.com/conan-io/examples2/pull/177/commits/1588288ccab4c1c09e875191a15ad2669fddf2e9)

memsharded reviewed (commented):
It might be better to not have a ``test_package()`` for it, but a regular other package recipe that uses it explicitly, so it is visible the ``if self.settings.os == "Linux": self.requires("ncurses/system")`` explicitly

uilianries commented:
> It might be better to not have a test_package() for it, but a regular other package recipe that uses it explicitly, so it is visible the if self.settings.os == "Linux": self.requires("ncurses/system") explicitly

Addressed on the commit [3874e44](https://github.com/conan-io/examples2/pull/177/commits/3874e44a54df6a503bc9da051e267ff65a10cdae)

Removed the test package, but added a "consumer" application instead. Also, adapted the CI strip to not only build both packages but also find and execute the new application.

memsharded commented on examples/tools/system/package_manager/ci_test_example.py:
```diff
@@ -0,0 +1,40 @@
+
+print("- Consuming Conan package ncurses/system -")
+
+out = run("conan create consumer/ --version=0.1.0 {}".format(" ".join(["-c " + conf for conf in confs])))
+
+assert "Conan: Target declared 'ncurses::ncurses'" in out
+assert "package(): Packaged 1 file: ncurses_version" in out
+
+match = re.search(r"Full package reference:\s*(.+)", out)
+assert match
+full_package_ref = match.group(1)
+
+package_folder = run(f"conan cache path {full_package_ref}").strip()
+assert os.path.isdir(package_folder)
+ncurses_version = os.path.join(package_folder, "bin", "ncurses_version")
+assert os.path.isfile(ncurses_version)
+
+out = run(ncurses_version)
```
I think this is a bit more complicated than it should:
- Use a simpler consumer recipe, no name, no version, no ``package_info()``
- Have a ``build()`` method that does the build and also ``self.run(..., env="conanrun")`` the executable
- call ``conan build .``

uilianries replied:
Done, only running `conan build` now in order to build the consumer: [fef2c93](https://github.com/conan-io/examples2/pull/177/commits/fef2c93cd2987861166ddcb7d9c1fca8c3e39831)

memsharded commented on examples/tools/system/package_manager/consumer/conanfile.py:
```diff
@@ -0,0 +1,32 @@
+        cmake.configure()
+        cmake.build()
+
+    def package(self):
+        cmake = CMake(self)
+        cmake.install()
+
+    def package_info(self):
+        self.cpp_info.libdirs = []
+        self.cpp_info.includedirs = []
```
Can be removed

memsharded reviewed (commented):
Almost there, looking good

memsharded reviewed (approved):
Great!

uilianries commented:
Add extra support for Mac (homebrew) and FreeBSD (pkg):

- https://www.freshports.org/devel/ncurses/
- https://formulae.brew.sh/formula/ncurses

uilianries commented:
The commit [f4dcb75](https://github.com/conan-io/examples2/pull/177/commits/f4dcb757f42f5bafbf94824a6c470dce61294716) improves the consumer example, by using screen feature of ncurses to shows a centralized message:

![Screenshot from 2025-02-03 14-04-50](https://github.com/user-attachments/assets/a8da471f-4966-4922-8d48-334295b57832)

The user only needs to press any key and it will go back to the terminal.

uilianries commented:
The commit simplified the consumer example: It shows a colored message on the screen, with ncurses version, then, exits:

![Screenshot from 2025-02-03 15-18-41](https://github.com/user-attachments/assets/0d510b2a-f48a-4a87-8f3d-6ba85a3ee80e)

uilianries commented on examples/tools/system/package_manager/consumer/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.15)
+project(ncurses_version C)
+
+find_package(ncurses CONFIG REQUIRED)
```
Not bad idea at all! https://cmake.org/cmake/help/latest/module/FindCurses.html

So we should use `Curses::Curses` as well for cmake_target_name

uilianries replied:
@czoido Done, please, check the commit e8e45a2

czoido commented on examples/tools/system/package_manager/consumer/conanfile.py:
```diff
@@ -0,0 +1,30 @@
+        cmake.configure()
+        cmake.build()
+
+        if self.settings.os in _supported_os:
```
```suggestion
        if self.settings.os in self._supported_os:
```

czoido commented on examples/tools/system/package_manager/consumer/conanfile.py:
```diff
@@ -0,0 +1,30 @@
+        return ["Linux", "Macos", "FreeBSD"]
+
+    def requirements(self):
+        if self.settings.os in _supported_os:
```
```suggestion
        if self.settings.os in self._supported_os:
```

memsharded commented:
Great, good job!

uilianries commented:
@memsharded @czoido I just skipped the example app execution on the commit [e3607c1](https://github.com/conan-io/examples2/pull/177/commits/e3607c1b757ce05b3a41e0b6940b5df39907c84c)

As we are using ncurses, it will need a configured terminal, which is not the case in the current CI that's building the examples. We could handle terminfo as workaround, but sounds too much. Locally, we should not have this problem when running the application.
